File: packages/ts/src/strategies/fix-newlines.ts
/**
 * fix-newlines: Replace literal newlines inside JSON string values with escaped \n.
 * Must run LAST (after all other structural repairs).
 */
export const fixNewlines = (text: string): string => {
  let result = "";
  let inString = false;
  let i = 0;

  while (i < text.length) {
    const ch = text[i];
    if (!inString) {
      if (ch === '"') inString = true;
      result += ch;
      i++;
      continue;
    }
    // Inside string
    if (ch === "\\") {
      // Consume escape pair
      result += ch;
      i++;
      if (i < text.length) { result += text[i]; i++; }
      continue;
    }
    if (ch === '"') {
      inString = false;
      result += ch;
      i++;
      continue;
    }
    if (ch === "\n") {
      result += "\\n";
      i++;
      continue;
    }
    if (ch === "\r") {
      result += "\\r";
      i++;
      continue;
    }
    result += ch;
    i++;
  }
  return result;
};

File: packages/python/src/output_guard/strategies/fix_newlines.py
"""fix-newlines: Replace literal newlines inside JSON string values with \\n.

Mirror of packages/ts/src/strategies/fix-newlines.ts. Runs LAST.
"""

from __future__ import annotations


def fix_newlines(text: str) -> str:
    out: list[str] = []
    in_string = False
    i = 0
    n = len(text)
    while i < n:
        ch = text[i]
        if not in_string:
            if ch == '"':
                in_string = True
            out.append(ch)
            i += 1
            continue
        if ch == "\\":
            out.append(ch)
            i += 1
            if i < n:
                out.append(text[i])
                i += 1
            continue
        if ch == '"':
            in_string = False
            out.append(ch)
            i += 1
            continue
        if ch == "\n":
            out.append("\\n")
            i += 1
            continue
        if ch == "\r":
            out.append("\\r")
            i += 1
            continue
        out.append(ch)
        i += 1
    return "".join(out)
